util/cobrautil/completion.go

```diff
@@ -0,0 +1,62 @@
+package cobrautil
+
+import (
+	"strings"
+
+	"github.com/docker/buildx/bake"
+	"github.com/docker/buildx/builder"
+	"github.com/docker/buildx/store/storeutil"
+	"github.com/docker/cli/cli/command"
+	"github.com/spf13/cobra"
+)
+
+// ValidArgsFn defines a completion func to be returned to fetch completion options
+type ValidArgsFn func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective)
+
+func NoCompletion(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	return nil, cobra.ShellCompDirectiveNoSpace
+}
+
+func CompleteBakeTargets(files []string) ValidArgsFn {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		f, err := bake.ReadLocalFiles(files)
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+		tgts, err := bake.ListTargets(f)
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveError
+		}
+		var filtered []string
+		if toComplete == "" {
+			return tgts, cobra.ShellCompDirectiveNoFileComp
+		}
+		for _, tgt := range tgts {
+			if strings.HasPrefix(tgt, toComplete) {
+				filtered = append(filtered, tgt)
+			}
```

Comment:
Could we split out the filtering logic into a separate function to share with other completion funcs as well?

Reply:
Don't think it's worth it currently. Also targets and builders are not of the same type so would need some reflection.

Reply:
Generics! 😂 

![Uploading magic.gif…]()